File: scripts/lib/popups.mjs
/**
 * 关页面弹窗（仅 dialog / popup / modal；串行执行，不碰 Escape / 全屏遮罩）
 */
import { ada, sleep } from "./ada.mjs";
import { WEB_DISMISS_DOM_CLICK_SCRIPT } from "./popups-dismiss-dom.mjs";
import {
  WEB_POPUP_BLOCKER_PROBE_SCRIPT,
  WEB_POPUP_IDLE_POLLS,
  WEB_POPUP_PRE_WAIT_POLL_MS
} from "./popups-wait-dom.mjs";

export { WEB_DISMISS_DOM_CLICK_SCRIPT } from "./popups-dismiss-dom.mjs";

export const DEFAULT_DISMISS_TIMEOUT_MS = 10_000;

const DOM_SCAN_BURST = 4;
const DISMISS_ACTION_WAIT_MS = 1_200;
/** 关弹窗探测：文案未命中时尽快跳过（默认 600 易拖长单轮） */
const DISMISS_LOCATOR_TIMEOUT_MS = 300;
const DISMISS_HIT_SLEEP_MS = 200;
const DISMISS_ROUND_SLEEP_MS = 200;
const POPUP_ROOT =
  '[role="dialog"],dialog,[class*="modal" i],[class*="popup" i],[aria-modal="true"],' +
  '[class*="login-layer" i],[class*="login-modal" i],[class*="login-popup" i],[class*="login-bottom-bar" i],' +
  '[id*="dialog-wrap" i],[id*="dialog" i]';

/** 仅在弹窗根节点内的关闭按钮 */
const WEB_DISMISS_LOCATORS = [
  {
    css:
      `${POPUP_ROOT} [id*="close" i], ${POPUP_ROOT} [class*="close" i], ` +
      `${POPUP_ROOT} [class*="cancel" i], ${POPUP_ROOT} [class*="dismiss" i]`
  },
  {
    css:
      `#login2025-dialog-wrap [id*="close" i], #login2025-dialog-wrap [class*="close" i], ` +
      `#login2025-dialog-wrap [aria-label*="关闭"], #login2025-dialog-wrap [aria-label*="close" i]`
  },
  {
    css:
      `${POPUP_ROOT} [aria-label*="关闭"], ${POPUP_ROOT} [aria-label*="close" i], ` +
      `${POPUP_ROOT} [title*="关闭"], ${POPUP_ROOT} [title*="close" i]`
  },
  {
    css:
      `${POPUP_ROOT} img[id*="close" i], ${POPUP_ROOT} img[class*="close" i], ` +
      `${POPUP_ROOT} button[id*="close" i], ${POPUP_ROOT} button[class*="close" i]`
  },
  {
    css:
      `${POPUP_ROOT} [id^="close" i], ${POPUP_ROOT} [id$="close" i], ` +
      `${POPUP_ROOT} [class^="close" i], ${POPUP_ROOT} [class$="close" i]`
  },
  { css: `${POPUP_ROOT} [data-dismiss="modal"]` }
];

const MOBILE_DISMISS_LABELS = [
  "关闭",
  "跳过",
  "我知道了",
  "知道了",
  "暂不",
  "不再提示",
  "取消",
  "×",
  "Close",
  "Got it"
];

/**
 * @param {number|false|null|undefined|{ timeoutMs?: number, attempts?: number }} arg
 *   不传 → 默认 {@link DEFAULT_DISMISS_TIMEOUT_MS}（10s）；数字为**整次调用总时长上限**（非每轮各 N 秒）
 * @param {number|undefined} attemptsArg 最多轮数（与 timeoutMs 同时生效，先到先停）
 */
export function normalizeDismissOpts(arg, attemptsArg) {
  const normalizeAttempts = (value) => {
    if (value == null) return Number.POSITIVE_INFINITY;
    const n = Number(value);
    if (!Number.isFinite(n)) return Number.POSITIVE_INFINITY;
    return Math.max(1, Math.floor(n));
  };

  if (arg === undefined || arg === null || arg === false) {
    return { timeoutMs: DEFAULT_DISMISS_TIMEOUT_MS, attempts: normalizeAttempts(attemptsArg) };
  }
  if (typeof arg === "number") {
    return { timeoutMs: Math.max(0, arg), attempts: normalizeAttempts(attemptsArg) };
  }
  return {
    timeoutMs: arg.timeoutMs ?? DEFAULT_DISMISS_TIMEOUT_MS,
    attempts: normalizeAttempts(arg.attempts ?? attemptsArg)
  };
}

function dismissPayload(options = {}) {
  const waitTimeoutMs =
    options.dismissWaitMs ?? options.dismissActionWaitMs ?? DISMISS_ACTION_WAIT_MS;
  return {
    ...options,
    waitTimeoutMs,
    /** 探测关弹窗：未命中返回 UI_ELEMENT_NOT_FOUND 业务码，不抛系统异常 */
    optional: true,
    bestEffort: true,
    locatorTimeoutMs: options.locatorTimeoutMs ?? DISMISS_LOCATOR_TIMEOUT_MS
  };
}

/** 关弹窗探测：任何底层异常都视为业务未命中，不向上抛 */
async function safeAda(platform, sessionId, command, payload) {
  try {
    return await ada(platform, sessionId, command, payload);
  } catch (error) {
    return {
      success: false,
      errorCode: "UI_ELEMENT_NOT_FOUND",
      errorMessage: error instanceof Error ? error.message : String(error),
      data: { businessCode: "LOCATOR_NOT_FOUND", optional: true }
    };
  }
}

/**
 * @typedef {object} DismissPopupsResult
 * @property {boolean} success
 * @property {boolean} dismissed
 * @property {string} businessCode POPUP_DISMISSED | POPUP_NOT_FOUND | POPUP_DISMISS_TIMEOUT
 * @property {string} reason dismissed | no_popup | timed_out
 * @property {number} dismissActions
 * @property {number} rounds
 * @property {boolean} timedOut
 * @property {number} elapsedMs
 * @property {number} timeoutMs
 * @property {string[]} hits
 */

async function waitForWebPopupReady(sessionId, payload, budgetMs) {
  const deadline = Date.now() + Math.max(0, budgetMs);
  let idleStreak = 0;
  let sawBlocker = false;
  while (Date.now() < deadline) {
    const r = await safeAda("web", sessionId, "custom", {
      ...payload,
      action: "evaluate",
      script: WEB_POPUP_BLOCKER_PROBE_SCRIPT
    });
    if (!r.success) break;
    const value = r.data?.value;
    if (value?.blocking) {
      sawBlocker = true;
      idleStreak = 0;
      return { ready: true, reason: "blocking", id: value.id ?? "blocker" };
    }
    idleStreak += 1;
    if (idleStreak >= WEB_POPUP_IDLE_POLLS) {
      return { ready: true, reason: sawBlocker ? "cleared" : "idle" };
    }
    await sleep(WEB_POPUP_PRE_WAIT_POLL_MS);
  }
  return { ready: true, reason: "timeout", sawBlocker };
}

export async function dismissWebPopups(sessionId, options = {}, dismissArg, attemptsArg) {
  const { timeoutMs, attempts } = normalizeDismissOpts(dismissArg, attemptsArg);
  const payload = dismissPayload(options);
  const started = Date.now();
  const deadline = started + timeoutMs;
  let dismissActions = 0;
  let rounds = 0;
  let idleStreak = 0;
  const hitLog = [];

  const preBudget = Math.min(4000, Math.max(600, Math.floor(timeoutMs * 0.45)));
  const pre = await waitForWebPopupReady(sessionId, payload, preBudget);
  if (pre.reason === "blocking") hitLog.push(`pre:${pre.id ?? "blocker"}`);
  else if (pre.reason === "idle") hitLog.push("pre:idle");

  while (Date.now() < deadline && rounds < attempts) {
    rounds += 1;
    let roundOk = false;

    for (let i = 0; i < DOM_SCAN_BURST; i++) {
      if (Date.now() >= deadline) break;
      const r = await safeAda("web", sessionId, "custom", {
        ...payload,
        action: "evaluate",
        script: WEB_DISMISS_DOM_CLICK_SCRIPT
      });
      if (!r.success) break;
      const value = r.data?.value;
      if (!value?.clicked) break;
      roundOk = true;
      hitLog.push(`dom:${value.via}:${value.text ?? value.tag ?? "?"}`);
      await sleep(DISMISS_HIT_SLEEP_MS);
      break;
    }

    if (!roundOk) {
      for (const locator of WEB_DISMISS_LOCATORS) {
        if (Date.now() >= deadline) break;
        const r = await safeAda("web", sessionId, "click", { ...payload, locator });
        if (!r.success) continue;
        roundOk = true;
        hitLog.push(`locator:${JSON.stringify(locator).slice(0, 72)}`);
        await sleep(DISMISS_HIT_SLEEP_MS);
        break;
      }
    }

    if (roundOk) {
      dismissActions += 1;
      idleStreak = 0;
    } else {
      idleStreak += 1;
      if (idleStreak >= 2) break;
    }
    if (Date.now() >= deadline) break;
    await sleep(DISMISS_ROUND_SLEEP_MS);
  }

  const endedAt = Date.now();
  const dismissed = dismissActions > 0;
  const timedOut = endedAt >= deadline;
  return {
    success: true,
    dismissed,
    businessCode: dismissed ? "POPUP_DISMISSED" : timedOut ? "POPUP_DISMISS_TIMEOUT" : "POPUP_NOT_FOUND",
    reason: dismissed ? "dismissed" : timedOut ? "timed_out" : "no_popup",
    dismissActions,
    rounds,
    timedOut,
    elapsedMs: endedAt - started,
    timeoutMs,
    hits: hitLog
  };
}

export async function dismissMobilePopups(platform, sessionId, base, screen, dismissArg, attemptsArg) {
  const { timeoutMs, attempts } = normalizeDismissOpts(dismissArg, attemptsArg);
  const payload = dismissPayload(base);
  const started = Date.now();
  const deadline = started + timeoutMs;
  let dismissActions = 0;
  let rounds = 0;
  let idleStreak = 0;
  const hitLog = [];

  while (Date.now() < deadline && rounds < attempts) {
    rounds += 1;
    let roundOk = false;

    for (const text of MOBILE_DISMISS_LABELS) {
      if (Date.now() >= deadline) break;
      const r = await safeAda(platform, sessionId, "click", {
        ...payload,
        locator: { text }
      });
      if (!r.success) continue;
      roundOk = true;
      hitLog.push(`text:${text}`);
      await sleep(DISMISS_HIT_SLEEP_MS);
      break;
    }

    if (!roundOk && Date.now() < deadline) {
      const x = Math.round((screen?.width ?? 1080) * 0.92);
      const y = Math.round((screen?.height ?? 2400) * 0.08);
      const corner = await safeAda(platform, sessionId, "click", { ...payload, point: [x, y] });
      if (corner.success) {
        roundOk = true;
        hitLog.push(`point:${x},${y}`);
      }
    }

    if (roundOk) {
      dismissActions += 1;
      idleStreak = 0;
    } else {
      idleStreak += 1;
      if (idleStreak >= 2) break;
    }
    if (Date.now() >= deadline) break;
    await sleep(DISMISS_ROUND_SLEEP_MS);
  }

  const endedAt = Date.now();
  const dismissed = dismissActions > 0;
  const timedOut = endedAt >= deadline;
  return {
    success: true,
    dismissed,
    businessCode: dismissed ? "POPUP_DISMISSED" : timedOut ? "POPUP_DISMISS_TIMEOUT" : "POPUP_NOT_FOUND",
    reason: dismissed ? "dismissed" : timedOut ? "timed_out" : "no_popup",
    dismissActions,
    rounds,
    timedOut,
    elapsedMs: endedAt - started,
    timeoutMs,
    hits: hitLog
  };
}

File: scripts/lib/popups.py
"""关页面弹窗（Web / 移动），与 popups.mjs 对齐。"""
from __future__ import annotations

import re
import time
from pathlib import Path
from typing import Any, Callable

DEFAULT_DISMISS_TIMEOUT_MS = 10_000
DOM_SCAN_BURST = 4
DISMISS_LOCATOR_TIMEOUT_MS = 300
DISMISS_HIT_SLEEP_S = 0.2
DISMISS_ROUND_SLEEP_S = 0.2

_POPUP_ROOT = (
    '[role="dialog"],dialog,[class*="modal" i],[class*="popup" i],[aria-modal="true"],'
    '[id*="dialog-wrap" i],[id*="dialog" i]'
)
WEB_DISMISS_LOCATORS: tuple[dict[str, Any], ...] = (
    {
        "css": (
            f'{_POPUP_ROOT} [id*="close" i], {_POPUP_ROOT} [class*="close" i], '
            f'{_POPUP_ROOT} [class*="cancel" i], {_POPUP_ROOT} [class*="dismiss" i]'
        )
    },
    {
        "css": (
            '#login2025-dialog-wrap [id*="close" i], '
            '#login2025-dialog-wrap [class*="close" i], '
            '#login2025-dialog-wrap [aria-label*="关闭"], '
            '#login2025-dialog-wrap [aria-label*="close" i]'
        )
    },
    {
        "css": (
            f'{_POPUP_ROOT} [aria-label*="关闭"], '
            f'{_POPUP_ROOT} [aria-label*="close" i], '
            f'{_POPUP_ROOT} [title*="关闭"], '
            f'{_POPUP_ROOT} [title*="close" i]'
        )
    },
    {
        "css": (
            f'{_POPUP_ROOT} img[id*="close" i], {_POPUP_ROOT} img[class*="close" i], '
            f'{_POPUP_ROOT} button[id*="close" i], {_POPUP_ROOT} button[class*="close" i]'
        )
    },
    {
        "css": (
            f'{_POPUP_ROOT} [id^="close" i], {_POPUP_ROOT} [id$="close" i], '
            f'{_POPUP_ROOT} [class^="close" i], {_POPUP_ROOT} [class$="close" i]'
        )
    },
    {"css": f'{_POPUP_ROOT} [data-dismiss="modal"]'},
)

MOBILE_DISMISS_LABELS: tuple[str, ...] = (
    "关闭",
    "跳过",
    "我知道了",
    "知道了",
    "暂不",
    "以后再说",
    "不再提示",
    "取消",
    "拒绝",
    "×",
    "Close",
    "Got it",
)


def _load_web_dismiss_dom_script() -> str:
    path = Path(__file__).resolve().parent / "popups-dismiss-dom.mjs"
    text = path.read_text(encoding="utf-8")
    m = re.search(r"export const WEB_DISMISS_DOM_CLICK_SCRIPT = `([\s\S]*?)`;", text)
    if not m:
        raise RuntimeError("WEB_DISMISS_DOM_CLICK_SCRIPT not found in popups-dismiss-dom.mjs")
    return m.group(1)


_WEB_DISMISS_DOM_SCRIPT = _load_web_dismiss_dom_script()


def _load_web_popup_probe_script() -> str:
    path = Path(__file__).resolve().parent / "popups-wait-dom.mjs"
    text = path.read_text(encoding="utf-8")
    m = re.search(r"export const WEB_POPUP_BLOCKER_PROBE_SCRIPT = `([\s\S]*?)`;", text)
    if not m:
        raise RuntimeError("WEB_POPUP_BLOCKER_PROBE_SCRIPT not found in popups-wait-dom.mjs")
    return m.group(1)


_WEB_POPUP_PROBE_SCRIPT = _load_web_popup_probe_script()
WEB_POPUP_PRE_WAIT_POLL_S = 0.2
WEB_POPUP_IDLE_POLLS = 2


def normalize_dismiss_opts(
    timeout_or_opts: int | dict[str, Any] | None = None, attempts: int | None = None
) -> tuple[int, int | None]:
    timeout_ms: int | None = None
    tries: int | None = attempts
    if isinstance(timeout_or_opts, dict):
        timeout_ms = timeout_or_opts.get("timeoutMs") or timeout_or_opts.get("timeout_ms")
        if tries is None:
            tries = timeout_or_opts.get("attempts")
    else:
        timeout_ms = timeout_or_opts
    timeout = DEFAULT_DISMISS_TIMEOUT_MS if timeout_ms is None else int(timeout_ms)
    max_attempts = max(1, int(tries)) if tries is not None else None
    return timeout, max_attempts


def dismiss_probe_payload(base: dict[str, Any]) -> dict[str, Any]:
    return {
        **base,
        "optional": True,
        "bestEffort": True,
        "locatorTimeoutMs": base.get("locatorTimeoutMs", DISMISS_LOCATOR_TIMEOUT_MS),
    }


def _ada(platform: str, session_id: str, command: str, payload: dict[str, Any]) -> dict[str, Any]:
    from ada_client import ada as ada_fn

    return ada_fn(platform, session_id, command, payload)


def _safe_ada(platform: str, session_id: str, command: str, payload: dict[str, Any]) -> dict[str, Any]:
    try:
        return _ada(platform, session_id, command, payload)
    except Exception as exc:
        return {
            "success": False,
            "errorCode": "UI_ELEMENT_NOT_FOUND",
            "errorMessage": str(exc),
            "data": {"businessCode": "LOCATOR_NOT_FOUND", "optional": True},
        }


def _dismiss_web_dom_scan_burst(session_id: str, options: dict[str, Any]) -> bool:
    for _ in range(DOM_SCAN_BURST):
        r = _safe_ada(
            "web",
            session_id,
            "custom",
            {**options, "action": "evaluate", "script": _WEB_DISMISS_DOM_SCRIPT},
        )
        if not r.get("success"):
            break
        value = (r.get("data") or {}).get("value") or {}
        if not value.get("clicked"):
            break
        time.sleep(DISMISS_HIT_SLEEP_S)
        return True
    return False


def _wait_for_web_popup_ready(session_id: str, options: dict[str, Any], budget_s: float) -> dict[str, Any]:
    deadline = time.time() + max(0.0, budget_s)
    idle_streak = 0
    saw_blocker = False
    while time.time() < deadline:
        r = _safe_ada(
            "web",
            session_id,
            "custom",
            {**options, "action": "evaluate", "script": _WEB_POPUP_PROBE_SCRIPT},
        )
        if not r.get("success"):
            break
        value = (r.get("data") or {}).get("value") or {}
        if value.get("blocking"):
            saw_blocker = True
            idle_streak = 0
            return {"ready": True, "reason": "blocking", "id": value.get("id") or "blocker"}
        idle_streak += 1
        if idle_streak >= WEB_POPUP_IDLE_POLLS:
            return {"ready": True, "reason": "cleared" if saw_blocker else "idle"}
        time.sleep(WEB_POPUP_PRE_WAIT_POLL_S)
    return {"ready": True, "reason": "timeout", "sawBlocker": saw_blocker}


def dismiss_web_popups(
    session_id: str,
    options: dict[str, Any],
    timeout_or_opts: int | dict[str, Any] | None = None,
    attempts: int | None = None,
) -> dict[str, Any]:
    timeout, max_attempts = normalize_dismiss_opts(timeout_or_opts, attempts)
    opts = {
        **options,
        "waitTimeoutMs": options.get("dismissWaitMs")
        or options.get("dismissActionWaitMs")
        or 1200,
    }
    started = time.time()
    deadline = started + timeout / 1000.0
    dismiss_actions = 0
    rounds = 0
    idle_streak = 0
    hit_log: list[str] = []

    pre_budget_s = min(4.0, max(0.6, (timeout / 1000.0) * 0.45))
    pre = _wait_for_web_popup_ready(session_id, opts, pre_budget_s)
    if pre.get("reason") == "blocking":
        hit_log.append(f"pre:{pre.get('id') or 'blocker'}")
    elif pre.get("reason") == "idle":
        hit_log.append("pre:idle")

    while time.time() < deadline and (max_attempts is None or rounds < max_attempts):
        rounds += 1
        round_ok = False
        if time.time() < deadline and _dismiss_web_dom_scan_burst(session_id, opts):
            round_ok = True
            hit_log.append("dom:scan")
        if not round_ok:
            for loc in WEB_DISMISS_LOCATORS:
                if time.time() >= deadline:
                    break
                r = _safe_ada("web", session_id, "click", {**opts, "locator": loc})
                if r.get("success"):
                    round_ok = True
                    hit_log.append(f"locator:{loc}")
                    time.sleep(DISMISS_HIT_SLEEP_S)
                    break
        if round_ok:
            dismiss_actions += 1
            idle_streak = 0
        else:
            idle_streak += 1
            if idle_streak >= 2:
                break
        if time.time() >= deadline:
            break
        time.sleep(DISMISS_ROUND_SLEEP_S)

    ended = time.time()
    elapsed_ms = int((ended - started) * 1000)
    return {
        "success": True,
        "dismissed": dismiss_actions > 0,
        "businessCode": "POPUP_DISMISSED"
        if dismiss_actions > 0
        else ("POPUP_DISMISS_TIMEOUT" if ended >= deadline else "POPUP_NOT_FOUND"),
        "reason": "dismissed"
        if dismiss_actions > 0
        else ("timed_out" if ended >= deadline else "no_popup"),
        "dismissActions": dismiss_actions,
        "rounds": rounds,
        "timedOut": ended >= deadline,
        "elapsedMs": elapsed_ms,
        "timeoutMs": timeout,
        "hits": hit_log,
    }


def _mobile_dismiss_round(
    platform: str,
    session_id: str,
    base: dict[str, Any],
    screen_w: int,
    screen_h: int,
    deadline: float,
    hit_log: list[str],
) -> bool:
    probe = dismiss_probe_payload(base)
    for label in MOBILE_DISMISS_LABELS:
        if time.time() >= deadline:
            break
        r = _safe_ada(platform, session_id, "click", {**probe, "locator": {"text": label}})
        if r.get("success"):
            hit_log.append(f"text:{label}")
            time.sleep(DISMISS_HIT_SLEEP_S)
            return True
    if time.time() >= deadline:
        return False
    corner = _safe_ada(
        platform,
        session_id,
        "click",
        {**probe, "point": [int(screen_w * 0.92), int(screen_h * 0.08)]},
    )
    if corner.get("success"):
        hit_log.append(f"point:{int(screen_w * 0.92)},{int(screen_h * 0.08)}")
        return True
    return False


def dismiss_mobile_popups(
    platform: str,
    session_id: str,
    base: dict[str, Any],
    screen_w: int,
    screen_h: int,
    timeout_or_opts: int | dict[str, Any] | None = None,
    attempts: int | None = None,
    *,
    on_round: Callable[[int, bool], None] | None = None,
) -> dict[str, Any]:
    timeout, max_attempts = normalize_dismiss_opts(timeout_or_opts, attempts)
    started = time.time()
    deadline = started + timeout / 1000.0
    dismiss_actions = 0
    rounds = 0
    idle_streak = 0
    hit_log: list[str] = []

    while time.time() < deadline and (max_attempts is None or rounds < max_attempts):
        rounds += 1
        round_ok = _mobile_dismiss_round(
            platform, session_id, base, screen_w, screen_h, deadline, hit_log
        )
        if on_round is not None:
            on_round(rounds, round_ok)
        if round_ok:
            dismiss_actions += 1
            idle_streak = 0
        else:
            idle_streak += 1
            if idle_streak >= 2:
                break
        if time.time() >= deadline:
            break
        time.sleep(DISMISS_ROUND_SLEEP_S)

    ended = time.time()
    elapsed_ms = int((ended - started) * 1000)
    return {
        "success": True,
        "dismissed": dismiss_actions > 0,
        "businessCode": "POPUP_DISMISSED"
        if dismiss_actions > 0
        else ("POPUP_DISMISS_TIMEOUT" if ended >= deadline else "POPUP_NOT_FOUND"),
        "reason": "dismissed"
        if dismiss_actions > 0
        else ("timed_out" if ended >= deadline else "no_popup"),
        "dismissActions": dismiss_actions,
        "rounds": rounds,
        "timedOut": ended >= deadline,
        "elapsedMs": elapsed_ms,
        "timeoutMs": timeout,
        "hits": hit_log,
    }
